casper/src/test/scala/coop/rchain/casper/util/rholang/DeployerAuthTest.scala

```diff
@@ -0,0 +1,117 @@
+import coop.rchain.models.rholang.implicits._
+import coop.rchain.models.{GDeployerAuth, Par}
+import coop.rchain.rholang.interpreter.accounting
+import monix.eval.Task
+import monix.execution.Scheduler.Implicits.global
+import org.scalatest.{FlatSpec, Matchers}
+
+import scala.concurrent.duration._
+
+class DeployerAuthTest extends FlatSpec with Matchers {
+
+  private val runtimeManager: Resource[Task, RuntimeManager[Task]] =
+    mkRuntimeManager("deployer-auth-runtime-manager-test")
+
+  "Deployer authentication" should "be forwarded to the reducer" in {
+    val sk = PrivateKey(
+      Base16.unsafeDecode("b18e1d0045995ec3d010c387ccfeb984d783af8fbb0f40fa7db126d889f6dadd")
+    )
+    val pk             = ByteString.copyFrom(Ed25519.toPublic(sk).bytes)
+    val captureChannel = "__DEPLOYER_AUTH_VALUE__"
+    val deploy = ConstructDeploy.sourceDeploy(
+      source = s"""new auth(`rho:deployer:auth`) in { @"$captureChannel"!(*auth) }""",
+      timestamp = System.currentTimeMillis(),
+      accounting.MAX_VALUE,
+      sec = sk
+    )
+    val result =
+      runtimeManager
+        .use(mgr => mgr.captureResults(mgr.emptyStateHash, deploy, captureChannel))
+        .runSyncUnsafe(10.seconds)
+
+    result.size should be(1)
+    result.head should be(GDeployerAuth(pk): Par)
+  }
+
+  it should "make drain vault attacks impossible" in {
+    val skDeployer = PrivateKey(
+      Base16.unsafeDecode("b18e1d0045995ec3d010c387ccfeb984d783af8fbb0f40fa7db126d889f6dadd")
+    )
+    val skAttacker = PrivateKey(
+      Base16.unsafeDecode("c18e1d0045995ec3d010c387ccfeb984d783af8fbb0f40fa7db126d889f6dadd")
+    )
+    val deploy = ConstructDeploy.sourceDeploy(
+      source =
+        s"""new auth(`rho:deployer:auth`) in { contract @"checkAuth"(input, ret) = { ret!(*input == *auth) }}""",
+      timestamp = System.currentTimeMillis(),
+      accounting.MAX_VALUE,
+      sec = skDeployer
+    )
+    val captureChannel = "__RETURN_VALUE__"
+    val checkAuth =
+      s"""new auth(`rho:deployer:auth`), ret in {
+         |@"checkAuth"!(*auth, *ret) |
+         |  for(isAuthenticated <- ret) {
+         |    @"$captureChannel"!(*isAuthenticated)
+         |  }
+         |} """.stripMargin
+    val attack = ConstructDeploy.sourceDeploy(
+      source = checkAuth,
+      timestamp = 0L,
+      accounting.MAX_VALUE,
+      sec = skAttacker
+    )
+    val nonAttack = ConstructDeploy.sourceDeploy(
+      source = checkAuth,
+      timestamp = 0L,
+      accounting.MAX_VALUE,
+      sec = skDeployer
+    )
+    val result =
+      runtimeManager
+        .use { mgr =>
+          mgr
+            .computeState(mgr.emptyStateHash)(Seq(deploy), 0L)
+            .flatMap { result =>
+              val hash = result._1
+              mgr
+                .captureResults(
+                  hash,
+                  attack,
+                  captureChannel
+                )
+            }
+        }
+        .runSyncUnsafe(10.seconds)
```

Comment:
Instead of organizing the code by technical aspect (source1, source2, compute1, compute2, assert1, assert2), please reorganize by domain aspect (sourceAttack, computeAttack, assertAttack, similar for non-attack) and then extract the duplication between the two blocks into a method. The whole test should be similar in length to:

```
checkAccessGranted(deployer, deployer, true)
checkAccessGranted(deployer, attacker, false)
```

Reply:
Done